Elm UI reset button

julianpistorius opened the pull request:
Partial refactoring to indicate how to use Elm UI widgets.

- Added a new reset button using the Elm UI widget library
- Shows how you can intermingle Html and Elm UI widgets

Next: 

- [x] Replace increment and decrement buttons (currently Html.button) with Elm UI buttons (Element.Input.button)
- [x] Replace the text label displaying the number (currently Html.text) with Elm UI text widget of some sort. 
- [x] Ensure none of the view is using the widgets from the Html package, and uses just Elm UI widgets. This means that the top level of the view function should look like:
```elm
view : Model -> Html Msg
view model =
    Element.layout [] ( 
      -- All Elm UI widgets here
     )
```

julianpistorius committed:
Added a new reset button.

This button is created by calling the Element.Input.button function from the
Elm UI widget library (as opposed to the HTML.button function used for
increment and decrement).

julianpistorius committed:
Added a Reset message, and made the new Elm UI reset button send it when pressed.

vsoch commented on src/Main.elm:
```diff
@@ -1,6 +1,8 @@
 
 import Browser
+import Element
+import Element.Input
```
okay, so these imports correspond to the UI library that I installed. If I hadn't installed them, this would not work. And like any library, it provides a namespace of functions (and elements?) that I can use in my code.

julianpistorius replied:
Exactly!

julianpistorius replied:
Have a look at the Elm UI package documentation for examples:

- https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/
- https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element

julianpistorius replied:
I'm going to clarify something. You asked: 

> And like any library, it provides a namespace of functions (and elements?) that I can use in my code.

There are no 'elements' in Elm. The `Element` above exists solely because @mdgriffith decided to name the top level module in the Elm UI package `Element` since it was based on his work in [mdgriffith/style-elements](https://github.com/mdgriffith/style-elements).

Elm has **functions** and **values**: https://guide.elm-lang.org/core_language.html

> The smallest building block in Elm is called a **value**. This includes things like `42`, `True`, and `"Hello!"`.

> A **function** is a way to transform values. Take in one value, and produce another.

Values as well as functions have **types**: https://guide.elm-lang.org/types/reading_types.html

Most of the types you'll come across are:

- Primitives (String, Int, Bool, Float)
- Lists (ordered collection of values)
- Records (a bit like a type-safe ~dictionary~ Python namedtuple  - e.g. a value like: `{ firstName = "Hermann", lastName = "Hesse" }` has type `{ firstName: String, lastName: String}`)
- Custom types (like the `Msg` type we define on line 29)

Elm libraries (such as Elm UI and Html) expose **functions** and **type declarations**.

That's about it. In Elm you solve problems by thinking of data flowing through functions which take values of some type and return values of some other types. 

You might think of it like the Unix pipeline with strict type-checking added.

I hope this helps.

vsoch commented on src/Main.elm:
```diff
@@ -24,7 +26,7 @@ init =
 
-type Msg = Increment | Decrement
+type Msg = Increment | Decrement | Reset
 
```
What is this line saying? Is Increment / Decrement something known that is imported, or are they being defined here? I'd normally read this like "the Msg type can be an Increment or a Decrement" but I didn't see where those are defined.

julianpistorius replied:
The latter. This line is defining a new 'custom type' in Elm lingo. Think of it like an enum but even cooler. 

My PR extended the custom type by adding a third option - 'Reset'

vsoch commented on src/Main.elm:
```diff
@@ -24,7 +26,7 @@ init =
+type Msg = Increment | Decrement | Reset
 
 update : Msg -> Model -> Model
 update msg model =
```
How would you read these lines? (29 and 30)

julianpistorius replied:
See: https://guide.elm-lang.org/types/custom_types.html

I'd read like this:

The Msg type has three **variants**. A Msg can be `Increment`, `Decrement`, or `Reset`

vsoch commented on src/Main.elm:
```diff
@@ -35,13 +37,22 @@ update msg model =
-    , button [ onClick Increment ] [ text "+" ]
-    ]
+    div []
+        [ button [ onClick Decrement ] [ text "-" ]
```
is the definition of Increment and Decrement down here, and Elm is okay with finding it later (after the definition as a type?)

julianpistorius replied:
No, the definition of Increment and Decrement is on line 29. Increment and Decrement have to exist in the namespace for the view function to be able to use them.

vsoch commented:
I'm loving this! I've never had someone teach me so clearly, and with a beautiful PR, and now I have my next set of TODOs. I will likely jump on this tomorrow, deep in Python atm. Thank you so much, this is the best!

julianpistorius commented:
My pleasure. Code as conversation FTW! 😸

vsoch committed:
Replacing increment and decrement buttons (currently Html.button) with Elm UI buttons (Element.Input.button) and adding background colors

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
okay I need help on the second point! I first tried replacing text with Element.text, but then I got an error about what div needs. So now I'm trying to get something that works with (Element.el? and using Element.text?) I keep getting different errors.

```
        , Element.layout []
        ,   (Element.el []
              Element.text (String.fromInt counter)
            )
```
There is also still a "div" at the top so maybe that's the issue?

```
This `layout` call produces:

    Element.Element msg -> Html msg

But all the previous elements in the list are:

    Html Action

Hint: Everything in a list must be the same type of value. This way, we never
run into unexpected values partway through a List.map, List.foldl, etc. Read
<https://elm-lang.org/0.19.1/custom-types> to learn how to “mix” types.
```

vsoch commented:
I just pushed the changes (and notes) that I made for the first bullet, so we are looking at the same thing :) Those changes don't have me mucking around with replacing Html.text with Element.text.

julianpistorius commented:
Close @vsoch! 

The hint says:

```
Hint: Everything in a list must be the same type of value.
```

Have a look at this example for how to make text label in with Elm UI (hint: `Element.text` needs to be wrapped in something like `Element.el`):

https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element#el

Something like this should work:

```diff
@@ -74,7 +74,9 @@
               , label = Element.text "-"
               }
             )
-        , div [] [ text (String.fromInt counter) ]
+        , Element.layout [] (
+            Element.el [] (Element.text (String.fromInt counter))
+        )
         , Element.layout []
             (Element.Input.button [
                 Background.color marigold
```

Then you will probably want to replace the `Html.div` with something like [`Element.column`](https://package.elm-lang.org/packages/mdgriffith/elm-ui/1.1.5/Element#column). See the type annotation for the `column` function, and how it takes a list of Elm-UI elements as its second parameter?

You should end up with something like this, where you've pulled the `Element.layout` function to the top of the `view` function:

```elm
view : Counter -> Html Action
view counter =
    Element.layout []
        (Element.column []
            [
                -- First Elm UI widget
                -- , Second Elm UI widget
                -- ... etc. 
            ]
        )
```

Here's an example:

https://github.com/mdgriffith/elm-ui/blob/1.1.5/examples/Form.elm#L83

Let me know how you get on.

julianpistorius commented on src/Main.elm:
```diff
@@ -2,55 +2,91 @@ module Main exposing (..)
 -- MODEL
 
-type alias Model = Int
+type alias Counter = Int
```
Renaming Model to Counter might make sense for a small app like this, but in any application where you have a multiple values in your model you'll probably want to stick with the 'Model' naming convention.

https://guide.elm-lang.org/architecture/buttons.html#model

> The point of the **model** is to capture all the details about your application as data.

vsoch replied:
okay I'll change this back! But for a new user, or someone just reading the code not familiar with Elm, it's very weird because the variable "Model" (intuitively) is describing a Counter, and Model better describes something that is for a database, view, or similar. The same for Msg that I will change from Action - in this app it isn't really a message, it's more like an action to be done. I suspect this is one of those "conventions" that I'm not allowed to argue with, that seems totally straight forward to old timers, but programmers like myself are like ??

julianpistorius replied:
Yes, I can see how that can be confusing. 

One way to refactor this to be clearer might be to use a record instead of an Int. Something like:

```elm
type alias Model =
    { counter : Int }
```

This way you can also start adding other values to the model. Let's imagine you wanted the user to be able to select the text background color - you might change your Model type to something like this:

```elm
type alias Model =
    { counter : Int
    , textBackground : Element.Color
    }
```

Your view function could use it like so:

```elm
            -- snip
            , Element.el [ Background.color model.textBackground ] (Element.text (String.fromInt model.counter))
            -- snip
```

vsoch replied:
I like this much better, because it's more clearly a message to the developer about what Model actually means. Trying this now!

julianpistorius commented on src/Main.elm:
```diff
@@ -2,55 +2,91 @@ module Main exposing (..)
 -- UPDATE
 
-type Msg = Increment | Decrement | Reset
+type Action = Increment | Decrement | Reset
```
Two reasons why this might not be a good idea:

1. The convention is to call this type `Msg`. Calling it `Action` will confuse other Elm programmers.
2. The semantics for how this type is used by the Elm runtime is more like a 'message' and less like an 'action'

Elaborating on the second point - 'Action' implies that your code is 'performing an action' (possibly with side-effects), which is a less accurate mental model for what happens.

See this description of the Elm Architecture: <https://guide.elm-lang.org/architecture/buttons.html#overview>

> Elm starts by rendering the initial value on screen. From there you enter into this loop:
> 
> 1. Wait for user input.
> 2. Send a message to update
> 3. Produce a new Model
> 4. Call view to get new HTML
> 5. Show the new HTML on screen
> 6. Repeat!

Later on you'll get to 'commands'. These are closer to the semantics of an action, except you're 'commanding' the Elm runtime to 'perform actions' on your behalf.

See here for more info: <https://guide.elm-lang.org/effects/>

vsoch replied:
This will solidify in my brain eventually. :) And you are right - I was close! I was off by a few braces.

julianpistorius commented on README.md:
```diff
@@ -226,9 +230,82 @@ Hey that seemed to work! Here is the updated [elm.json](elm.json):
+akin to looking up functions and usage for any module. Things that I experimented
+with include:
+
+ - changing a variable named "Model" to "Counter" because it represents an integer for the counter number, and I didn't think "model" described that accurately.
```
Note: Both `Model` and `Msg` are types, not variables. 

There are no variables in Elm. It really takes immutability seriously. You can only define constants, and only once in a namespace.

Example:

```elm
update : Action -> Counter -> Counter
update action counter =
    let
        value1 =
            "Some string"

        value2 =
            "Some other string"

        value1 =
            "Changing value 1"
    in
    case action of
        -- snip
```

This will give you a compiler error:

```
elm make src/Main.elm 
Detected problems in 1 module.
-- NAME CLASH ----------------------------------------------------- src/Main.elm

This `let` expression defines `value1` more than once! One here:

77|         value1 =
            ^^^^^^
And another one here:

71|         value1 =
            ^^^^^^
How can I know which one you want? Rename one of them!
```

vsoch committed:
Replacing the text label displaying the number (currently Html.text) with Elm UI text widgets (el and column).

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch committed:
removing div from import

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
@julianpistorius I think that was the final changes we needed! I want to clarify that we are still using Html from the Html package (it matches now the view format that we would expect). Does everything look good here? If yes, we can merge, and then I would like to ask for what you'd suggest I tackle next (broad description or to whatever degree you think appropriate is good with me, I'll ask questions if/when needed). It hasn't clicked yet, but it will eventually, always does :)

vsoch committed:
updating model to be a record

this way I can have an Int inside the record for a counter,
and this is more intutitive for the developer to read and know
we are talking about a counter.

Signed-off-by: vsoch <vsochat@stanford.edu>

julianpistorius commented:
@vsoch Yes this is looks good!

The only thing I would suggest before merging is to run `elm-analyse` and `elm-format` on the code. It's like linting for Elm.

- <https://github.com/stil4m/elm-analyse>
- <https://github.com/avh4/elm-format>

You'll see that it picks up a few things like unused imports:

```
$ elm-analyse
...
Found 6 message(s)

Messages:
- src/Main.elm
  > Exposing all at ((1,23),(1,25))
  > Duplicate import for module `Element`` at [ ((4,1),(4,15)) | ((5,1),(5,33)) ]
  > Unused imported variable `onClick` at ((9,30),(9,37))
  > No signature for top level definition `marigold` at ((13,1),(13,9))
  > No signature for top level definition `green` at ((16,1),(16,6))
  > No signature for top level definition `main` at ((21,1),(21,5))
```

```
$ elm-format --validate .
[{"path":"./src/Main.elm","message":"File is not formatted with elm-format-0.8.3 --elm-version=0.19"}
]
$ elm-format src/Main.elm
This will overwrite the following files to use Elm's preferred style:

    src/Main.elm

This cannot be undone! Make sure to back up these files before proceeding.

Are you sure you want to overwrite these files with formatted versions? (y/n)

y
Processing file src/Main.elm
$ elm-format --validate .
[]
```

Note: I have my IDE set to run `elm-format` automatically when I save an Elm file.

> I want to clarify that we are still using Html from the Html package

Yes, we only need it for the view function type signature though. If we remove the import you would get this compiler error:

```elm
$ elm make src/Main.elm 
Detected problems in 1 module.
-- NAMING ERROR --------------------------------------------------- src/Main.elm

I cannot find a `Html` type:

87| view : Model -> Html Msg
```

You ask:

> ... then I would like to ask for what you'd suggest I tackle next

I think it's time to get familiar with JSON decoding, as well 'effects' (particularly making HTTP requests). Play around with these two examples:

- <https://guide.elm-lang.org/effects/http.html>
- <https://guide.elm-lang.org/effects/json.html>

Try the following:

- Change the views to use Elm-UI functions instead of Html functions
- Modify the JSON example to load the list of SCIF apps from this URL: <https://sci-f.github.io/apps/api/index.json>

The following resources on JSON decoders will help:

- <https://github.com/zwilias/elm-demystify-decoders>
- <https://thoughtbot.com/blog/5-common-json-decoders>
- <https://thoughtbot.com/blog/getting-unstuck-with-elm-json-decoders>

Once you've written a few simple JSON decoders by hand you might want to look at some of the Elm code generators:

- <https://github.com/sporto/awesome-elm#code-generators>

vsoch commented:
okay, all set! I do like this much better. I also updated the UI so the increment is on top.

julianpistorius committed:
Fixed some problems reported by `elm-analyse`:

  > Exposing all at ((1,23),(1,25))
  > Duplicate import for module `Element`` at [ ((4,1),(4,15)) | ((5,1),(5,33)) ]
  > Unused imported variable `onClick` at ((9,30),(9,37))
  > No signature for top level definition `marigold` at ((13,1),(13,9))
  > No signature for top level definition `green` at ((16,1),(16,6))
  > No signature for top level definition `main` at ((21,1),(21,5))

julianpistorius committed:
Ran `elm-format` on Main.elm

vsoch commented:
I'm adding it to install to the container now - hang tight!

vsoch committed:
adding elm-analyse and elm-format to container for linting

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
okay, added to Dockerfile! https://github.com/vsoch/elm-app-catalog/pull/1/commits/becc984c24f19f7a0b3901756aa4e7ee98b2caf4 If this looks good, we can merge here, and I'll start poking around / working on what you mentioned above.

julianpistorius commented:
LGTM! Merge it.

vsoch commented:
Woohoo!! 

I'll try to make some time over the weekend / next week for next steps.
